doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+the host for the schema version of the specification itself, which may
+be defined in a third package entirely.
+
+Consider :
+```python
+#Rough psuedocode
+specVersion = hostInterface.getSchemaVersion(CompanyCorpCamera.getPackageIdentifier())
+switch (specVersion):
+  case 1:
+    handleTraitVersionsForSpecV1(successCallback, errorCallback)
+  case 2:
+    handleTraitVersionsForSpecV2(successCallback, errorCallback)
+
+def handleTraitVersionsForSpecV1(successCallback, errorCallback):
+  cameraTraitVersion = hostInterface.getSchemaVersion(legacy.CompanyCorpCameraV1.Camera.getPackageIdentifier())
+  companyCorpTraitVersion = hostInterface.getSchemaVersion(legacy.CompanyCorpCameraV1.CompanyCorp.getPackageIdentifier())
+
+  # You still have to version iterate over each trait individually, because you can't assume they came from the same package.
+  switch (cameraTraitVersion):
+    case 1:
+      ... # Camera trait version 1 behaviour
+    case 2:
+      ... # Camera trait version 2 behaviour
+
+  switch (companyCorpTraitVersion):
+  case 1:
+    ... # CompanyCorp trait version 1 behaviour
+  case 2:
+    ... # CompanyCorp trait version 2 behaviour
+
+def handleTraitVersionsForSpecV2:
+  ... #Same again but slightly different spec structure.
+```
+
+As you can see, this isn't going to be a good time. Which is a shame
+because in all other cases except for split-package specifications, a
+single schema version has seemed the simplest option.
+
+The main problem however is the generated specification views have
+utility methods on them to create the trait views that they should
+contain. This introduces a sticky build dependency that would need to be
+removed, likely by removing the utility methods overall.
+
+## Simplify, bundle everything.
+A solution that came up when trying to wrap our heads round this was the
+idea of not allowing cross-package specifications as such, but to
+instead achieve composition via bundled packages.
+
+In this conception, when you generate your trait and spec views, you
+would provide the `.yaml` of all the other packages you are referencing,
+essentially duplicating them into a new, unique package.
+
+`openassetio-traitgen -o output -g python --deps
+openassetio-mediacreation/traits.yml companycorp_traits.yml`
+
+This would essentially create a brand new traits package with it's own
+unique trait identifier `companycorp-openassetio-mediacreation`, which
+could then be treated as a single package, sidestepping _all_ the
+complexity without removing the ability to compose cross package traits.
+
+
+This would strongly put multi-package specifications into the camp of
+being an extension method used by organizations wishing to use
+openassetio inside their private pipeline, as it's doubtful productized
+management systems or hosts would want to account for all the specific
+permutations here. A major downside.
+
+## Simplify, ignore it.
+Also an option, maybe we can think about this later. How and if external
```

Comment:
This feels like the best option to me.  For OAIO v1 release, are non-medicreation traits a required feature?  I still think we should be versioning mediacreation traits, but perhaps we can defer third-party supplied traits until the library has seen some real world use.

Reply:
The intention has always been that host applications will supply their own application-specific Traits and Specifications.  E.g. a `BlenderWorkfileSpecification` that is bundled with Blender and extends the traits of `WorkfileSpecification` with a `Blender` trait, allowing managers to filter queries to only show Blender project files.

The idea is that such application-specific traits/specs will be made public by ISVs so that manager plugins can, if they wish, adapt their code to support them.

Reply:
Even given that, that dosen't neccesarily mean non-mediacreation _specifications_ are a required feature. We've always expected mixin traits to be appended to formal specifications, if there's only a few examples of mixin traits to expect (like the host identifying on, for example,) we might be able to move forward asking managers to just check for those if they need em.